pkg/migration/migration.go

```diff
@@ -0,0 +1,259 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+// Package migration captures the facilities needed to define and execute
+// migrations for a crdb cluster. These migrations can be arbitrarily long
+// running, are free to send out arbitrary requests cluster wide, change
+// internal DB state, and much more. They're typically reserved for crdb
+// internal operations and state. Each migration is idempotent in nature, is
+// associated with a specific cluster version, and executed when the cluster
+// version is made active on every node in the cluster.
+//
+// Examples of migrations that apply would be migrations to move all raft state
+// from one storage engine to another, or purging all usage of the replicated
+// truncated state in KV.
+package migration
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/clusterversion"
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/security"
+	"github.com/cockroachdb/cockroach/pkg/server/serverpb"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/logtags"
+)
+
+// Manager coordinates long-running migrations.
+type Manager interface {
+	RunDependencies
+	Migrate(ctx context.Context, user security.SQLUsername, from, to clusterversion.ClusterVersion) error
+}
+
+// Registry provides access to migrations.
+type Registry interface {
+	GetMigration(key clusterversion.ClusterVersion) (Migration, bool)
+}
+
+// RunDependencies are used by the job to run migrations.
+type RunDependencies interface {
```

Comment:
Does "RunDependencies" match with terminology we use elsewhere in the jobs infrastructure? Or is the "Run" here referring the runnable migrations (KVMigrations.Run, SQLMigrations.Run)? I don't have a better name off-hand, though compared to the rest of the interfaces we've defined here, this was a bit less clear.

Reply:
I agree this was a wart. I've reworked it to be `migration.JobDeps` and pulled the `Registry` to just be defined inline. I think it's cleaner.